docs: add blog posts for npm-on-ipfs (#215)
ipfs/blog: content/post/70-putting-npm-on-ipfs-part-1.md

---
date: 2019-02-17

> title: Putting npm on IPFS Part 1 - The Registry

Reviewer: I like the structure, separating this into the server/client posts definitely makes it more digestible, and the breakout "quick how to" at the top of each is good 👍

author: Alex Potsides
---

> npm is the de facto package manager for the JavaScript ecosystem and is largest registry in the world, with more than [900k](https://replicate.npmjs.com/_all_docs) packages and over 7 billion downloads a week. In recent years it's got a lot faster and more reliable, but what if we were to take it to the next level and put it on the distributed web?

Reviewer 2: ```suggestion
npm is the de facto package manager for the JavaScript ecosystem and is largest registry in the world, with more than [900k](https://replicate.npmjs.com/_all_docs) packages and over 7 billion downloads a week. Nowadays npm is incredibly fast and reliable thanks to the hard work put by the NPM, Inc team, however we couldn't stop ourselves from wondering, what if we put the registry on the distributed web and to take it to a whole new level?
```

> npm is the de facto package manager for the JavaScript ecosystem and is largest registry in the world, with more than [900k](https://replicate.npmjs.com/_all_docs) packages and over 7 billion downloads a week. In recent years it's got a lot faster and more reliable, but what if we were to take it to the next level and put it on the distributed web?

Reviewer 3: It'd might be good to explain _why_ putting it on the distributed web takes it to the next level, what benefits does it bring to compared to using the npm registry?

Author: Done!

> npm is the de facto package manager for the JavaScript ecosystem and is largest registry in the world, with more than [900k](https://replicate.npmjs.com/_all_docs) packages and over 7 billion downloads a week. In recent years it's got a lot faster and more reliable, but what if we were to take it to the next level and put it on the distributed web?

Reviewer 4: ```suggestion
npm is the de facto package manager for the JavaScript ecosystem and the largest registry in the world, with more than [900k](https://replicate.npmjs.com/_all_docs) packages and over 7 billion downloads a week. In recent years it's got a lot faster and more reliable, but what if we were to take it to the next level and put it on the distributed web?
```

Author: I've made this change.

> 

Reviewer 2: Suggestion: add another paragraph to the intro that describes that this is Part I of a two part blog post series and share the outlines of each blog post

Author: Done!

## 🗄️ registry.js.ipfs.io

> 

Reviewer 2: Needs a Diagram explaining how to collector pulls the data and stores to IPFS

Author: I've added this in the Updates section.

We're maintaining a complete mirror of npm at https://registry.js.ipfs.io - the difference being it adds [CID](https://docs.ipfs.io/guides/concepts/cid)s to the metadata for every package which you can use to fetch packages from IPFS!

You can use the registry today by specifying the `--registry` parameter to npm:

```console
$ npm --registry=https://registry.js.ipfs.io install
```

...or Yarn:

```console
$ yarn --registry=https://registry.js.ipfs.io
```

This will instruct it to use our registry mirror instead of the default registry.  You can add this to your npm config via:

```console
$ npm config set registry https://registry.js.ipfs.io
```

### 📦 GET your dependencies, proxy everything else

Only `GET` requests are honoured by our mirror, all other requests are forwarded on to the npm registry, so you if you publish while using our mirror, your package will still end up on the public registry.  Requests/response content is not logged so your credentials are never at rest on our servers.

Everything is also served over https unlike some very old npm modules which are served over http by the npm registry.

## 👷‍♀️ How we built it

> A group of [ipfs-npm-registry-mirror](https://github.com/ipfs-shipyard/ipfs-npm-registry-mirror) nodes are running, they contain a js-IPFS instance used to share and resolve modules and are connected to the wider IPFS network.  They also have a small http server used to respond to requests for module metadata and tarballs from the from the npm cli.  These all sit behind an http load balancer to distribute traffic to the mirrors.

Reviewer 3: ```suggestion
A group of [ipfs-npm-registry-mirror](https://github.com/ipfs-shipyard/ipfs-npm-registry-mirror) nodes are running, they contain a [js-IPFS](https://github.com/ipfs/js-ipfs) instance used to share and resolve modules and are connected to the wider IPFS network.  They also have a small http server used to respond to requests for module metadata and tarballs from the from the npm cli.  These all sit behind an http load balancer to distribute traffic to the mirrors.
```

Author: I've made this change.

![Network topology](/70-putting-npm-on-ipfs-part-1/network-topology.png)

> ### 🗃️ The datastore

Reviewer 5: Perhaps we could just change the language a bit and add a call to action for people interested in s3. If you were just skimming the article you might not realize the datastore is using s3.

----

### 🗃️ Backed by S3

Each mirror has an [IPFS Repository](https://github.com/ipfs/specs/tree/master/repo) that is backed by AWS S3 via an instance of [datastore-s3](https://github.com/ipfs/js-datastore-s3) - this lets us deploy the service as a set of immutable containers, handle huge amounts of data and scale up and down as required at the cost of a small amount of latency on transfers. 

If you'd like to leverage S3 for your IPFS node, check out the example of how to configure IPFS to use S3 in the [datastore-s3 examples folder](https://github.com/ipfs/js-datastore-s3/tree/v0.2.3/examples/full-s3-repo).

Reviewer 2: Also, here you can give another version of the diagram that shows that the store is actually on S3 to make it crazy cheap

Author: Done!
